smolyay/test_function_class.py

```diff
@@ -0,0 +1,6478 @@
+import abc
+
+import numpy
+
+class test_fun(abc.ABC):
+    """Abstract test function class
+
+    Stores data about each test function including call, name, the number of
+    dimensions, the domain, and how to normalize that domain to the interval
+    [-1,1]
+
+    Parameters
+    ----------
+    name : str
+        the name of the class
+
+    dim : int
+        the number of variables used as input
+
+    lower_bounds : list of floats of size 1xdim
+        the lower bounds of the domain of each variable
+
+    upper_bounds : list of floats of size 1xdim
+        the upper bounds of the domain of each variable
+
+    domain : list of lists
+        the domain of the inputs for each dimension
+    """
+    def __call__(self,x):
+        """Evaluate the test function
+
+        Parameters
+        ----------
+        x : list of floats
+            the input to the call method
+
+        Raises
+        ------
+        ValueError
+            if the input is outside is outside the domain
```

Comment:
Docstrings should be meaningful of what the parameters represent, not things like "input to the method" (that is obvious because it is an argument).
```suggestion
        """Evaluate the function.

        Parameters
        ----------
        x : list
            Function input.

        Raises
        ------
        ValueError
            If the input is outside the function domain.
```